File: day10/main.go
// Advent of Code 2023, Day 10
//
// AK, 10 Dec 2023

package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
)

var lines [][]byte

type Point struct {
	x, y int
}

// Shortest distance to any point on path
var dist map[Point]int

// For Part 2, keep track of any position that has ever been visited
var everVisited map[Point]bool
var seq map[Point]int

func main() {

	// Read the input file (must have no empty lines at end)
	fname := "sample.txt"
	fname = "sample2a.txt"
	//fname = "input.txt"
	data, _ := ioutil.ReadFile(fname)
	lines = bytes.Split(data, []byte("\n"))

	// Find starting position
	var start Point
	for y := 0; y < len(lines); y++ {
		for x := 0; x < len(lines[0]); x++ {
			if lines[y][x] == 'S' {
				start = Point{x, y}
				break
			}
		}
	}
	fmt.Println("Start at", start)

	// Follow all paths from start, collect distances
	dist = map[Point]int{}         // Initialize shortest distance to any point on path
	visited := []Point{}           // Points already visited on this traversal
	everVisited = map[Point]bool{} // For part 2
	seq = map[Point]int{}
	explore(start, 0, visited) // Explore from start, get distances

	// Part 1: Find the maximum distance from start
	ans1 := 0
	for _, v := range dist {
		if v > ans1 {
			ans1 = v
		}
	}
	fmt.Println("Part 1:", ans1) // 6927

	// Part 2: count up points that are enclosed by path
	ans2 := 0
	for y := 0; y < len(lines); y++ {
		for x := 0; x < len(lines[0]); x++ {
			p := Point{x, y}
			if !everVisited[p] && enclosed2(p) {
				ans2++
			}
		}
	}
	fmt.Println("Part 2 for", fname, ":", ans2) // Samples s/b 4, 8, 10
	//fmt.Println(seq)
}

// Explore all paths from this point, mark points as visited, and update
// global dictionary with shortest distances to each point
func explore(here Point, steps int, visited []Point) {

	// Mark this point as visited
	visited = append(visited, here)
	everVisited[here] = true
	_, ok := seq[here]
	if !ok {
		seq[here] = steps + 1
	}

	// Update distance to this point if shorter
	if dist[here] == 0 || steps < dist[here] {
		dist[here] = steps
	}

	// Explore each point from here that has not been explored
	for _, next := range nextSteps(here) {
		if !in(next, visited) { //visited[next] == 0 {
			explore(next, steps+1, visited) // pointer to visited?
		}
	}
}

// Return a list of the viable next steps from this position.
// | is a vertical pipe connecting north and south.
// - is a horizontal pipe connecting east and west.
// L is a 90-degree bend connecting north and east.
// J is a 90-degree bend connecting north and west.
// 7 is a 90-degree bend connecting south and west.
// F is a 90-degree bend connecting south and east.
func nextSteps(p Point) []Point {
	res := []Point{}

	// Above: bar, F, 7
	p1 := Point{p.x, p.y - 1}
	c := at(p1)
	if c == '|' || c == 'F' || c == '7' {
		res = append(res, p1)
	}

	// Below: vert bar, L, J
	p1 = Point{p.x, p.y + 1}
	c = at(p1)
	if c == '|' || c == 'L' || c == 'J' {
		res = append(res, p1)
	}

	// Left: dash, F, L
	p1 = Point{p.x - 1, p.y}
	c = at(p1)
	if c == '-' || c == 'F' || c == 'L' {
		res = append(res, p1)
	}

	// Right: dash, 7, J
	p1 = Point{p.x + 1, p.y}
	c = at(p1)
	if c == '-' || c == '7' || c == 'J' {
		res = append(res, p1)
	}

	return res
}

// Ray Casting algorithm, which involves casting a ray from the point in
// question in any direction and counting the number of times the ray intersects
// with the edges of the shape. If the number of intersections is odd, the point
// is inside the shape; if it is even, the point is outside the shape.
func enclosed2(p Point) bool {

	var vcuts, hcuts int

	// Count intersections to left or right
	// MAY BE FAILING BECAUSE IT TREATS CONTIGUOUS POINTS AS DIFFERENT, EVEN
	// IF THEY ARE ON THE SAME HORIZONTAL PATH. BUT YOU CAN'T JUST LOOK AT
	// THE POINTS, NEED TO CONSIDER IF TWO CONTIGUOUS POINTS WERE DRAWN NEXT
	// TO EACH OTHER, OR DURING A LOOP. CAN DO THIS BY KEEPING TRACK OF SEQUENCE
	// POINTS ARE VISITED, THEY ARE CONTIGUOUS IF ABS(DIFF) == 1.
	if p.x > 0 {
		hcuts = countIntersectionsWithPath(p, -1, 0)
	} else {
		hcuts = countIntersectionsWithPath(p, 1, 0)
	}

	// Count intersections above or below
	if p.y > 0 {
		vcuts = countIntersectionsWithPath(p, 0, -1)
	} else {
		vcuts = countIntersectionsWithPath(p, 0, 1)
	}

	// Return true if odd number of intersections
	fmt.Println(p, string(at(p)), hcuts, vcuts, hcuts%2, vcuts%2)
	return hcuts%2 != 0 //&& vcuts%2 != 0
}

func countIntersectionsWithPath(p Point, dx, dy int) int {

	// Move to the next position in given direction
	prev := p //Point{-1, -1}
	p1 := Point{p.x + dx, p.y + dy}
	var cuts int
	for at(p1) != 0 { // continue while within range

		// If we are on the path, increment counter,
		// unless previous point was also on path, and was drawn just before or after it
		if everVisited[p1] {
			if everVisited[prev] && abs(seq[p1]-seq[prev]) != 1 {
				cuts++
			}
		}
		prev = p1

		// Move to the next position in given direction
		p1 = Point{p1.x + dx, p1.y + dy}
	}

	return cuts
}

// Character at a point, 0 if out of bounds
func at(p Point) byte {
	if p.y < 0 || p.y >= len(lines) || p.x < 0 || p.x >= len(lines[0]) {
		return 0
	} else {
		return lines[p.y][p.x]
	}
}

// Is point in a list of points? (would be faster to use a map)
func in(p Point, pp []Point) bool {
	for _, x := range pp {
		if x == p {
			return true
		}
	}
	return false
}

// Absolute value
func abs(a int) int {
	if a < 0 {
		return -a
	} else {
		return a
	}
}

File: day10/main.rs
// Advent of Code 2023, Day 10
//
// Given a 2D terrain of symbols, start at 'S' and follow the chain of "pipe"
// shapes back to S. For Part 1, calculate the distance of the furthest-away
// shape, distance measured in either direction from start. For Part 2,
// count up how many positions are *not* inside the shape.
//
// AK, Rust solution 12 Oct 2024, more efficient algorithm than the Go
// version, which assumed there could be multiple entry points from any
// location, and used recursion to explore all possible paths.

use std::fs;
use std::collections::HashMap;

// A point on the map, just x,y coordinates. The "derive"
// stuff at the top gets the compiler to create traits to
// copy, compare, or hash a Point object.
#[derive(Copy, Clone, Eq, PartialEq, Hash)]
struct Point {
    x: usize,
    y: usize
}

fn main() {

    // Read input file into a list of rows, each a list of chars
    let fname = "input.txt";
    let map = read_char_matrix(fname);

	// Part 1: starting at 'S', explore the pipes until you get back to 'S',
	// recording the distance along the way.

	// Find the starting point, where 'S' is
    let start = find_char_in_matrix('S', &map);
    println!("Starting at row {}, col {}", start.y, start.x);

	// Dictionary of distances from start, also indicates whether a point has been visited
	let mut dists = HashMap::<Point,i32>::new();
	dists.insert(start, 0); // mark start as visited

	// Explore from here, until you get back to 'S'
	let mut dist = 0;  // current distance from start
	let mut p = start; // the current point
	loop {

        // Find the possible locations you could go from here (should only be one or two, as there
        // are no T-junctions)
		let opts = next_steps(&p, &map);
		if opts.len() > 2 {  // should never happen
			println!("More than 2 options!");
			break;
		}

        // Move to the unvisited option from this point; if no more unvisited locations, we're done
        // (don't need to explictely check if we're back at start)
		p = opts[0];
		if dists.contains_key(&p) && opts.len() > 1 { // i.e., already visited
			p = opts[1]; // take the other unvisited one
		}
		if dists.contains_key(&p) {
			println!("No more unvisited points to explore");
			break;
		}

		// Set the distance of the new point
		dist += 1;
		dists.insert(p, dist);
	}

	// Part 1 is the longest distance from start, equal to half of total distance traversed
	println!("Part 1 (s/b 6927) = {}", (dist + 1) / 2);

    // Part 2: Use ray tracing method on rows only, to count up points that are enclosed by path.
    // Involves traversing each row, and counting up how many times we cross the outline of our
    // path. Where the count is odd, we are inside the shape, otherwise outside.
	// Source: https://www.quora.com/What-are-the-algorithms-for-determining-if-a-point-is-inside-an-arbitrary-closed-shape-or-not
	let mut ans2 = 0;
	for y in 0..map.len()  {        // each row
		let mut bars = 0;           // number of borders encountered on this row
		for x in 0..map.len()  {    // each character on row
			let p = Point{x: x, y: y};
			if dists.contains_key(&p) && one_of(at(&p, &map), "S|LJ") {
				bars += 1;          // we just crossed another border
			} else if !dists.contains_key(&p) && bars % 2 == 1 {
				ans2 += 1;          // odd number of borders crossed, inside shape
			}
		}
	}

	println!("Part 2 (s/b 467) = {}", ans2);
}

// Return a list of the viable next steps from this position, depends on the
// current and next shape being compatible connectors.
// | is a vertical pipe connecting north and south.
// - is a horizontal pipe connecting east and west.
// L is a 90-degree bend connecting north and east.
// J is a 90-degree bend connecting north and west.
// 7 is a 90-degree bend connecting south and west.
// F is a 90-degree bend connecting south and east.
fn next_steps(p: &Point, map: &Vec<Vec<char>>) -> Vec<Point> {

	// For list of possible moves
	let mut res = Vec::<Point>::new();
  
 	// character at this position
	let c0 = at(p, map);     

	// Up: bar, F, 7
	if let Some(p1) = make_point(&p, 0, -1, map) {
		let c = at(&p1, map);
		if one_of(c0, "S|LJ") && one_of(c, "|F7") {
			res.push(p1);
		}
	}

	// Down: vert bar, L, J
	if let Some(p1) = make_point(&p, 0, 1, map) {
		let c = at(&p1, map);
		if one_of(c0, "S|F7") && one_of(c, "|LJ") {
			res.push(p1);
		}
	}	

	// Left: dash, F, L
	if let Some(p1) = make_point(&p, -1, 0, map) {
		let c = at(&p1, map);
		if one_of(c0, "S-J7") && one_of(c, "-FL") {
			res.push(p1);
		}
	}

	// Right: dash, 7, J
	if let Some(p1) = make_point(&p, 1, 0, map) {
		let c = at(&p1, map);
		if one_of(c0, "S-FL") && one_of(c, "-7J") {
			res.push(p1);
		}
	}

	// Return list of points
	res
}

// Make a new point based on original point, x and y adjusted,
// return None if out of bounds. 
// Warning: this may fail if dx or dy are < -1
// This is absolutely horrible, so much easier in Go or Python, where
// it's just one line of code, e.g., p = Point{x+2, y+1}
fn make_point(p: &Point, dx: i32, dy: i32, map: &Vec<Vec<char>>) -> Option<Point> {

    // Convert everything to signed integers, because usize won't allow adding 
    // integer, what an abomination is Rust
	let px: i32 = p.x.try_into().unwrap();
	let py: i32 = p.y.try_into().unwrap();
	let nrows: i32 = map.len().try_into().unwrap();
	let ncols: i32 = map[0].len().try_into().unwrap();

	// New coordinates, check for out of bounds, return None if so
	let x1 = px + dx;
	let y1 = py + dy;
	if y1 < 0 || y1 >= nrows || x1 < 0 || x1 >= ncols { // y out of bounds
		return None;
	} 

	// Create new point, need to convert back to unsigned
	let new_x: usize = (px + dx).try_into().unwrap();
	let new_y: usize = (py + dy).try_into().unwrap();
	Some(Point{x: new_x, y: new_y})
}

// Character at a point, 0 if out of bounds
// Note that we don't need to check for < 0 since coordinates are usize
fn at(p: &Point, map: &Vec<Vec<char>>) -> char {
	if  p.y >= map.len() || p.x >= map[0].len() {
		return '\0';
	} else {
		return map[p.y][p.x];
	}
}

// Check if a char is in string, could be done inline, but matches Go version
fn one_of(c: char, s: &str) -> bool {
	s.contains(c)
}

// Find a character in a "matrix", returns (0,0) if not found (but this 
// may be the valid answer, so check). Only used to find the initial 'S'.
fn find_char_in_matrix(ch: char, map: &Vec<Vec<char>>) -> Point {
    let mut r = 0;
    let mut c = 0;
    for ri in 0..map.len() {
        for ci in 0..map[ri].len() {
            if map[ri][ci] == ch {
                r = ri;
                c = ci;
                break;
            }
        }
        if map[r][c] == 'S' {
            break;
        }
    }
    Point{x: c, y: r}
}

// Read a text file as a matrix of chars, i.e., a list of rows, each a list of chars
fn read_char_matrix(fname: &str) -> Vec<Vec<char>> {
    let data = fs::read_to_string(fname).expect("Read error");
    let mut map = Vec::<Vec<char>>::new();
    for l in data.lines() {
        if l.len() > 0 {
            map.push(l.chars().collect());
        }
    }
    map
}
